pkg/serviceaccount/externaljwt/metrics/metrics.go

```diff
@@ -0,0 +1,152 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"crypto/sha256"
+	"errors"
+	"sync"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const (
+	namespace = "apiserver"
+	subsystem = "externaljwt"
+)
+
+var (
+	failedKeyFetch = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "fetch_jwks_error_total",
+			Help:           "Total failed attempts at syncing supported JWKs",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"grpc_status_code"},
+	)
+
+	totalKeyFetch = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "fetch_jwks_request_total",
+			Help:           "Total attempts at syncing supported JWKs",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"status"},
+	)
+
+	lastKeyFetchTimeStamp = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "jwks_last_fetch_time",
+			Help:           "Timestamp of the last successful or failed automatic reload of encryption configuration split by apiserver identity.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		nil,
+	)
+
+	tokenGenFailure = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "sign_jwt_error_total",
+			Help:           "Total failed attempts at signing JWT",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"grpc_status_code"},
+	)
+
+	tokenGenReqTotal = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "sign_jwt_request_total",
+			Help:           "Total attempts at signing JWT",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"status"},
+	)
+)
+
+var registerMetrics sync.Once
+var hashPool *sync.Pool
+
+func RegisterMetrics() {
+	registerMetrics.Do(func() {
+		hashPool = &sync.Pool{
+			New: func() interface{} {
+				return sha256.New()
+			},
+		}
+		legacyregistry.MustRegister(failedKeyFetch)
+		legacyregistry.MustRegister(totalKeyFetch)
+		legacyregistry.MustRegister(lastKeyFetchTimeStamp)
+		legacyregistry.MustRegister(tokenGenFailure)
+		legacyregistry.MustRegister(tokenGenReqTotal)
+	})
+}
+
+func RecordKeysFetchFailure(err error) {
+	failedKeyFetch.WithLabelValues(getErrorCode(err)).Inc()
+	totalKeyFetch.WithLabelValues("failed").Inc()
+}
+
+func RecordKeysFetchSuccess() {
+	totalKeyFetch.WithLabelValues("success").Inc()
+	recordKeyFetchTimeStamp()
+}
+
+func recordKeyFetchTimeStamp() {
+	lastKeyFetchTimeStamp.WithLabelValues().SetToCurrentTime()
```

Comment:
I thought this was going to record the data timestamp of the last fetch

Reply:
This un-exported func is only called when plugin tries to report a successful request count for Fetch(); So, current time is good.